File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/app/AppWithOptionalComponents.java
package ru.tinkoff.kora.kora.app.annotation.processor.app;

import ru.tinkoff.kora.application.graph.ValueOf;
import ru.tinkoff.kora.common.KoraApp;
import ru.tinkoff.kora.common.annotation.Root;

import javax.annotation.Nullable;
import java.util.Optional;

@KoraApp
public interface AppWithOptionalComponents {
    default PresentInGraph presentInGraph() {
        return new PresentInGraph();
    }

    @Root
    default NotEmptyOptionalParameter notEmptyOptionalParameter(Optional<PresentInGraph> param) {
        return new NotEmptyOptionalParameter(param.orElse(null));
    }

    @Root
    default EmptyOptionalParameter emptyOptionalParameter(Optional<NotPresentInGraph> param) {
        return new EmptyOptionalParameter(param.orElse(null));
    }

    @Root
    default NotEmptyValueOfOptional notEmptyValueOfOptional(ValueOf<Optional<PresentInGraph>> param) {
        return new NotEmptyValueOfOptional(param.get().orElse(null));
    }

    @Root
    default EmptyValueOfOptional emptyValueOfOptional(ValueOf<Optional<NotPresentInGraph>> param) {
        return new EmptyValueOfOptional(param.get().orElse(null));
    }

    @Root
    default NotEmptyNullable notEmptyNullable(@Nullable PresentInGraph param) {
        return new NotEmptyNullable(param);
    }

    @Root
    default EmptyNullable emptyNullable(@Nullable NotPresentInGraph param) {
        return new EmptyNullable(param);
    }

    class NotPresentInGraph {}

    class PresentInGraph {}

    record NotEmptyOptionalParameter(@Nullable PresentInGraph value) {}

    record EmptyOptionalParameter(@Nullable NotPresentInGraph value) {}

    record NotEmptyValueOfOptional(@Nullable PresentInGraph value) {}

    record EmptyValueOfOptional(@Nullable NotPresentInGraph value) {}

    record NotEmptyNullable(@Nullable PresentInGraph value) {}

    record EmptyNullable(@Nullable NotPresentInGraph value) {}

}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/app/AppWithOptionalComponents.kt
package ru.tinkoff.kora.kora.app.ksp.app

import ru.tinkoff.kora.application.graph.ValueOf
import ru.tinkoff.kora.common.KoraApp
import ru.tinkoff.kora.common.annotation.Root
import java.util.*

@KoraApp
interface AppWithOptionalComponents {
    fun presentInGraph(): PresentInGraph {
        return PresentInGraph()
    }

    @Root
    fun notEmptyOptionalParameter(param: Optional<PresentInGraph>): NotEmptyOptionalParameter {
        return NotEmptyOptionalParameter(param.orElse(null))
    }

    @Root
    fun emptyOptionalParameter(param: Optional<NotPresentInGraph>): EmptyOptionalParameter {
        return EmptyOptionalParameter(param.orElse(null))
    }

    @Root
    fun notEmptyValueOfOptional(param: ValueOf<Optional<PresentInGraph>>): NotEmptyValueOfOptional {
        return NotEmptyValueOfOptional(param.get().orElse(null))
    }

    @Root
    fun emptyValueOfOptional(param: ValueOf<Optional<NotPresentInGraph>>): EmptyValueOfOptional {
        return EmptyValueOfOptional(param.get().orElse(null))
    }

    @Root
    fun notEmptyNullable(param: PresentInGraph?): NotEmptyNullable {
        return NotEmptyNullable(param)
    }

    @Root
    fun emptyNullable(param: NotPresentInGraph?): EmptyNullable {
        return EmptyNullable(param)
    }

    class NotPresentInGraph
    class PresentInGraph
    data class NotEmptyOptionalParameter(val value: PresentInGraph?)

    data class EmptyOptionalParameter(val value: NotPresentInGraph?)

    data class NotEmptyValueOfOptional(val value: PresentInGraph?)

    data class EmptyValueOfOptional(val value: NotPresentInGraph?)

    data class NotEmptyNullable(val value: PresentInGraph?)

    data class EmptyNullable(val value: NotPresentInGraph?)
}
